app/models/shipping_rate.rb

```diff
@@ -0,0 +1,61 @@
+require 'active_shipping'
+
+class ShippingRate
+  include ActiveModel::Validations
+
+  attr_accessor :origin, :destination, :package
+
+  validates :origin, :destination, :package, presence: true
+  validate :valid_locations
+  validate :valid_package
+
+  def initialize(attributes = {})
+    attributes.each do |name, value|
+      send("#{name}=", value)
+    end
+  end
+
+  def ups_rates # TODO: put in actual credentials
+    ups = ActiveShipping::UPS.new(login: ENV["ACTIVESHIPPING_UPS_LOGIN"], password: ENV["ACTIVESHIPPING_UPS_PASSWORD"], key: ENV["ACTIVESHIPPING_UPS_KEY"])
+    format_ups_rates(get_rates_from_shipper(ups))
+  end
+
+  def usps_rates
+    usps = ActiveShipping::USPS.new(login: ENV['ACTIVESHIPPING_USPS_LOGIN'])
+    format_usps_rates(get_rates_from_shipper(usps))
+  end
+
+  private
+
+  def format_ups_rates(rates)
+    rates.sort_by(&:price).map do |rate|
+      { service: rate.service_name, price: rate.total_price, delivery_date: rate.delivery_date }
+    end
+  end
+
+  def get_rates_from_shipper(shipper)
+    response = shipper.find_rates(origin, destination, package)
+    response.rates
+  end
+
+  def format_usps_rates(rates)
+    rates.sort_by(&:price).map{ |rate|
+      { service: rate.service_name, price: rate.price, delivery_date: rate.delivery_date }
+    }
+  end
+
+  def valid_locations # OPTIMIZE: it might be fun to make these error messages be more descriptive / include location errors
+    unless self.origin.is_a?(ShippingLocation) && self.origin.valid?
+      errors = self.origin.is_a?(ShippingLocation) ? self.origin.errors : "It is not an instance of ShippingLocation."
```

Comment:
I don't often see rubyists validating the type of an object and I think that's because it make it impossible to duck-type a replacement object. Duck-typing is a core power feature of ruby so it's nice to keep it available. This is especially true if you are testing this and want to insert a mock.